File: src/dotnet/Harness.Engine/HarnessConfig.cs
using System.Text.Json;

namespace Harness.Engine;

/// <summary>
/// Fixed harness variables, externalized into a <c>harness.json</c> at the repo root.
/// Previously these were hardcoded constants scattered around (<see cref="TaskRegistry"/>,
/// <see cref="DocsReader"/>, <c>RefinementTasks</c>); centralizing them here lets each
/// flow/environment adjust the ceilings without recompiling. Missing or unreadable → falls
/// back to <see cref="Default"/> (same tolerance as <see cref="StateStore.LoadFrom"/>:
/// config is optional input, it can't bring down the run).
///
/// Top-level type (not nested) so it's servable by System.Text.Json's source generator, a
/// Native AOT requirement.
/// </summary>
public record HarnessConfig(
    int MaxSteps,
    int MaxInstructionChars,
    int DocsMaxChars,
    string DocsFolder,
    int TimeoutMs,
    string ContextResetMode,
    double ContextResetThreshold,
    int ContextFallbackFeatures)
{
    // Step ceiling: prevents an infinite loop that would burn tokens indefinitely.
    // MaxInstructionChars = 0 disables the cost ceiling (only the step one applies).
    // TimeoutMs is always enabled: a workspace config may tune it but cannot turn off the
    // per-step time guard.
    public static HarnessConfig Default { get; } = new(
        MaxSteps: 12,
        MaxInstructionChars: 0,
        DocsMaxChars: 40_000,
        DocsFolder: "specs",
        TimeoutMs: 10 * 60_000,
        ContextResetMode: "adaptive",
        ContextResetThreshold: 0.70,
        ContextFallbackFeatures: 1);

    private const string FilePath = "harness.json";

    // Hard ceiling on timeoutMs, regardless of the source (harness.json OR the env var
    // below). harness.json lives in the working directory the supervised agent itself
    // controls: without this ceiling, the agent could edit the file to grant itself an
    // arbitrarily high timeout and never get cut off by the time guard (see TaskRegistry).
    private const int MaxAllowedTimeoutMs = 10 * 60_000;
    private const int MinEnabledTimeoutMs = 1;

    // When set, overrides harness.json's timeoutMs. Unlike the file, the env var is set by
    // the parent process that invokes each harness step — outside the working directory
    // the supervised agent controls — so it can't be self-edited by the same agent the
    // timeout is meant to contain.
    private const string TimeoutMsEnvVar = "HARNESS_TIMEOUT_MS";

    // Loaded once per process (each harness invocation is a new process, so "once" = "per
    // loop turn"). Static readers — DocsReader, RefinementTasks — consume from here without
    // needing to receive the config as a parameter.
    private static HarnessConfig? _current;
    public static HarnessConfig Current => _current ??= Load();

    /// <summary>Forces a re-read of <c>harness.json</c> — for tests and long-lived drivers.</summary>
    public static void Reload() => _current = Load();

    /// <summary>Re-reads <c>harness.json</c> from disk; any failure returns <see cref="Default"/>.</summary>
    public static HarnessConfig Load()
    {
        var config = Default;
        try
        {
            var path = PathResolver.Resolve(FilePath);
            if (File.Exists(path))
            {
                var json = File.ReadAllText(path);
                var parsed = JsonSerializer.Deserialize(json, HarnessJsonContext.Default.HarnessConfig);
                if (parsed is not null)
                    config = parsed;
            }
        }
        catch (Exception ex)
        {
            HarnessLog.Error($"[HarnessConfig] failed to load; using defaults: {ex.Message}");
            config = Default;
        }

        return Normalize(ApplyTimeoutEnvOverride(config));
    }

    // See TimeoutMsEnvVar. Missing/invalid is silently ignored — same tolerance as the
    // rest of the config: it's optional input, it can't bring down the run.
    private static HarnessConfig ApplyTimeoutEnvOverride(HarnessConfig config)
    {
        var raw = Environment.GetEnvironmentVariable(TimeoutMsEnvVar);
        return int.TryParse(raw, out var timeoutMs) ? config with { TimeoutMs = timeoutMs } : config;
    }

    // A partial harness.json deserializes missing fields as 0/null. Zero is only valid
    // where it means "disabled" (cost ceilings); elsewhere, a missing field = default.
    private static HarnessConfig Normalize(HarnessConfig config) => config with
    {
        MaxSteps = config.MaxSteps > 0 ? config.MaxSteps : Default.MaxSteps,
        MaxInstructionChars = int.Max(config.MaxInstructionChars, 0),
        DocsMaxChars = config.DocsMaxChars > 0 ? config.DocsMaxChars : Default.DocsMaxChars,
        DocsFolder = string.IsNullOrWhiteSpace(config.DocsFolder) ? Default.DocsFolder : config.DocsFolder,
        // A workspace-controlled config may tune the timeout, but it may not turn the
        // guard off. The terminal latch in TaskRegistry is the second stop mechanism for
        // a timeout that already occurred.
        TimeoutMs = int.Clamp(
            config.TimeoutMs > 0 ? config.TimeoutMs : Default.TimeoutMs,
            MinEnabledTimeoutMs,
            MaxAllowedTimeoutMs),
        ContextResetMode = config.ContextResetMode is "adaptive" or "per-feature" or "never"
            ? config.ContextResetMode
            : Default.ContextResetMode,
        ContextResetThreshold = double.Clamp(
            config.ContextResetThreshold is > 0 and <= 1 ? config.ContextResetThreshold : Default.ContextResetThreshold,
            0.1,
            1.0),
        ContextFallbackFeatures = config.ContextFallbackFeatures > 0
            ? config.ContextFallbackFeatures
            : Default.ContextFallbackFeatures,
    };
}

File: src/go/harnessengine/harness_config.go
package harnessengine

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
)

// HarnessConfig holds the harness's fixed variables, externalized into a harness.json at
// the repo root. Centralizing them here lets each flow/environment tune the ceilings
// without recompiling. Absent or unreadable → falls back to DefaultHarnessConfig (same
// tolerance as the other stores: config is optional input, it must never bring the run down).
type HarnessConfig struct {
	MaxSteps                int     `json:"maxSteps"`
	MaxInstructionChars     int     `json:"maxInstructionChars"`
	DocsMaxChars            int     `json:"docsMaxChars"`
	DocsFolder              string  `json:"docsFolder"`
	TimeoutMs               int     `json:"timeoutMs"`
	ContextResetMode        string  `json:"contextResetMode"`
	ContextResetThreshold   float64 `json:"contextResetThreshold"`
	ContextFallbackFeatures int     `json:"contextFallbackFeatures"`
}

const harnessConfigPath = "harness.json"

// maxAllowedTimeoutMs is the hard upper bound on TimeoutMs, regardless of source (harness.json
// OR the env var below). harness.json lives in the working directory the supervised agent
// itself controls: without this ceiling, the agent could edit the file to grant itself an
// arbitrarily high timeout and never be cut by the time guard (see TaskRegistry).
const maxAllowedTimeoutMs = 5 * 60_000
const minEnabledTimeoutMs = 1

// timeoutMsEnvVar, when set, overrides harness.json's timeoutMs. Unlike the file, the env
// var is set by the parent process invoking each harness step — outside the working
// directory the supervised agent controls — so it cannot be self-edited by the same agent
// the timeout is meant to contain.
const timeoutMsEnvVar = "HARNESS_TIMEOUT_MS"

// DefaultHarnessConfig returns the built-in defaults. MaxInstructionChars = 0 disables the
// cost ceiling (only the step ceiling applies). TimeoutMs is always enabled: a workspace
// config may tune it but cannot turn off the time guard.
func DefaultHarnessConfig() HarnessConfig {
	return HarnessConfig{
		MaxSteps:                12,
		MaxInstructionChars:     0,
		DocsMaxChars:            40_000,
		DocsFolder:              "docs",
		TimeoutMs:               30_000,
		ContextResetMode:        "adaptive",
		ContextResetThreshold:   0.70,
		ContextFallbackFeatures: 1,
	}
}

var (
	configMu     sync.Mutex
	cachedConfig *HarnessConfig
)

// CurrentConfig is loaded once per process (each harness invocation is a fresh process, so
// "once" means "once per loop turn"). Static readers — DocsReader, RefinementTasks —
// consume from here without needing the config passed as a parameter.
func CurrentConfig() HarnessConfig {
	configMu.Lock()
	defer configMu.Unlock()
	if cachedConfig == nil {
		loaded := LoadConfig()
		cachedConfig = &loaded
	}
	return *cachedConfig
}

// ReloadConfig forces a re-read of harness.json — for tests and long-lived drivers.
func ReloadConfig() HarnessConfig {
	configMu.Lock()
	defer configMu.Unlock()
	loaded := LoadConfig()
	cachedConfig = &loaded
	return loaded
}

// LoadConfig re-reads harness.json from disk; any failure returns DefaultHarnessConfig.
func LoadConfig() HarnessConfig {
	config := DefaultHarnessConfig()

	path := ResolvePath(harnessConfigPath)
	if fileExists(path) {
		data, err := os.ReadFile(path)
		if err == nil {
			var parsed HarnessConfig
			if err := json.Unmarshal(data, &parsed); err == nil {
				config = parsed
			} else {
				fmt.Fprintf(os.Stderr, "[HarnessConfig] failed to load; using defaults: %s\n", err)
				config = DefaultHarnessConfig()
			}
		} else {
			fmt.Fprintf(os.Stderr, "[HarnessConfig] failed to load; using defaults: %s\n", err)
		}
	}

	return normalizeConfig(applyTimeoutEnvOverride(config))
}

// applyTimeoutEnvOverride: see timeoutMsEnvVar. Absent/invalid is silently ignored — same
// tolerance as the rest of the config: it's optional input, it can't bring the run down.
func applyTimeoutEnvOverride(config HarnessConfig) HarnessConfig {
	raw, ok := os.LookupEnv(timeoutMsEnvVar)
	if !ok {
		return config
	}
	timeoutMs, err := strconv.Atoi(strings.TrimSpace(raw))
	if err != nil {
		return config
	}
	config.TimeoutMs = timeoutMs
	return config
}

// normalizeConfig: a partial harness.json deserializes missing fields as 0/"". Zero is only
// valid where it means "disabled" (cost ceilings); elsewhere, a missing field means default.
func normalizeConfig(config HarnessConfig) HarnessConfig {
	def := DefaultHarnessConfig()

	if config.MaxSteps <= 0 {
		config.MaxSteps = def.MaxSteps
	}
	if config.MaxInstructionChars < 0 {
		config.MaxInstructionChars = 0
	}
	if config.DocsMaxChars <= 0 {
		config.DocsMaxChars = def.DocsMaxChars
	}
	if strings.TrimSpace(config.DocsFolder) == "" {
		config.DocsFolder = def.DocsFolder
	}
	if config.TimeoutMs <= 0 {
		config.TimeoutMs = def.TimeoutMs
	}
	config.TimeoutMs = clamp(config.TimeoutMs, minEnabledTimeoutMs, maxAllowedTimeoutMs)
	mode := strings.ToLower(strings.TrimSpace(config.ContextResetMode))
	if mode != "adaptive" && mode != "per-feature" && mode != "never" {
		config.ContextResetMode = def.ContextResetMode
	} else {
		config.ContextResetMode = mode
	}
	if config.ContextResetThreshold <= 0 {
		config.ContextResetThreshold = def.ContextResetThreshold
	}
	if config.ContextResetThreshold > 1 {
		config.ContextResetThreshold = 1
	}
	if config.ContextFallbackFeatures <= 0 {
		config.ContextFallbackFeatures = def.ContextFallbackFeatures
	}

	return config
}

func clamp(value, min, max int) int {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}
